File: api/app/services/refund_policy.py
"""Refund policy for a renter cancelling a *confirmed* reservation.

Tiered, matching the cancellation terms already shown to renters in the
booking UI (`web/src/components/ListingDetailModal.tsx`) and mirrored
client-side in `web/src/lib/refundPolicy.ts`:

    cancel >= 72h before the start date  -> 100% refund
    cancel  < 72h before the start date  ->  50% refund
    cancel on/after the start date       -> no self-serve cancellation

The refund is a fraction of the whole reservation total, Spacio's service
fee included — the fee comes back at the same rate as everything else, so a
50% refund also returns half the fee. On the Stripe side a single partial
`Refund` with `refund_application_fee=True` + `reverse_transfer=True` makes
Stripe apply that same proportion to the platform fee and to the transfer
already sent to the host (see
`services.reservation_payments.refund_reservation_payment`).

Pure and side-effect-free so it can be unit tested directly; the router
(`reservations.cancel_reservation`) turns a `RefundDecision` into the Stripe
call and the status write.
"""

from dataclasses import dataclass
from datetime import datetime

# Hours before the start date up to which a cancellation is still "free"
# (full refund). Inside this window the renter gets PARTIAL_REFUND_RATE.
FULL_REFUND_CUTOFF_HOURS = 72
PARTIAL_REFUND_RATE = 0.5


class CancellationNotAllowedError(Exception):
    """The reservation has already started — no self-serve cancellation."""


@dataclass(frozen=True)
class RefundDecision:
    tier: str  # "full" | "partial"
    refund_rate: float  # 1.0 | 0.5
    refund_amount: float  # dollars, rounded to whole cents
    refund_cents: int  # what to pass to stripe.Refund.create(amount=...)
    reason: str


def decide_refund(total_price: float, start: datetime, now: datetime) -> RefundDecision:
    """Return how much of ``total_price`` to refund for a cancellation at
    ``now`` against a reservation starting at ``start``. Raises
    ``CancellationNotAllowedError`` once the start date has passed."""
    if now >= start:
        raise CancellationNotAllowedError(
            "This reservation has already started. Contact support to cancel."
        )

    hours_until_start = (start - now).total_seconds() / 3600
    if hours_until_start >= FULL_REFUND_CUTOFF_HOURS:
        rate = 1.0
        tier = "full"
        reason = f"Cancelled {FULL_REFUND_CUTOFF_HOURS}+ hours before the start date."
    else:
        rate = PARTIAL_REFUND_RATE
        tier = "partial"
        reason = (
            f"Cancelled less than {FULL_REFUND_CUTOFF_HOURS} hours before the start date."
        )

    cents = round(total_price * rate * 100)
    return RefundDecision(
        tier=tier,
        refund_rate=rate,
        refund_amount=cents / 100,
        refund_cents=cents,
        reason=reason,
    )

File: web/src/lib/refundPolicy.ts
// Client-side mirror of api/app/services/refund_policy.py, used only to
// tell the renter what refund to expect *before* they confirm a
// cancellation. The server recomputes and is the source of truth. Keep the
// two tiers in sync with the backend if the policy changes.

export const FULL_REFUND_CUTOFF_HOURS = 72;
export const PARTIAL_REFUND_RATE = 0.5;

export type RefundEstimate = {
  tier: "full" | "partial" | "none";
  refundAmount: number;
  message: string;
};

// `startDate` is a date-only string ("2026-06-01"); the backend treats it
// as UTC midnight, so parse it the same way rather than in local time.
export function estimateRefund(startDate: string, totalPrice: number, now: Date = new Date()): RefundEstimate {
  const start = new Date(`${startDate}T00:00:00Z`);
  const hoursUntilStart = (start.getTime() - now.getTime()) / 3_600_000;

  if (hoursUntilStart <= 0) {
    return {
      tier: "none",
      refundAmount: 0,
      message: "This reservation has already started — contact support to cancel.",
    };
  }
  if (hoursUntilStart >= FULL_REFUND_CUTOFF_HOURS) {
    return {
      tier: "full",
      refundAmount: totalPrice,
      message: `You'll be refunded the full $${totalPrice.toFixed(2)}.`,
    };
  }
  const refundAmount = Math.round(totalPrice * PARTIAL_REFUND_RATE * 100) / 100;
  return {
    tier: "partial",
    refundAmount,
    message: `It's less than ${FULL_REFUND_CUTOFF_HOURS} hours before your start date, so you'll be refunded 50% ($${refundAmount.toFixed(
      2,
    )}).`,
  };
}
